README.md
Sample Onion Services
====================

Copyright (C) 2016-2017  F. Brezo and Y. Rubio, i3visio

1 - Description
---------------

This project includes a couple of onion services deployed with Bottle and Tor's stem.
By doing so, we want to provide an easy-to-use approach that makes to understand the
behaviour of certain Tor Gateways.


2 - License: GPLv3+
-------------------

This is free software, and you are welcome to redistribute it under certain conditions.

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.


For more details on this issue, check the [COPYING](COPYING) file.

3 - Installation
----------------

You will also need to have Tor up and running with several special specifications
Open the `/et/tor/torrc` file and search the `ControlPort 9051` line and uncomment
it. You will also need to uncomment the `CookieAuthentication 1` line to let `stem`
control the instance from Python code. Full details can be found in the 
[INSTALL.md](INSTALL.MD) file. Afterwards, restart tor service:
```
service tor restart
```

If everything went Ok, you can download the project knowing that you need a 
Python 2 installation (tested with 2.7) and pip. Check it using:
```
python --version
pip --version
```

You can now clone the project using `git` for example.
```
git clone https://github.com/i3visio/sample-onion-servers
```

Afterwards, fix the dependencies:
```
cd sample-onion-servers
pip install -r requirements.txt
```

You should be ready to go now.

4 - Basic usage
---------------

You should explore the `server` folder. There you will be able to find a couple 
of servers that may help you:
* `hello-onion.py`. Just a hello world as an example.
* `headers-onion.py`. It shows the headers sent by the user who performed the
request as seen by the server. Useful to test the queries sent by third parties
gateways without messing up too much.

5 - AUTHORS
-----------

More details about the authors in the [AUTHORS.md](AUTHORS.md) file.


CHANGES.md
Changelog
=========

For more information, check the README.md.

0.1.1, 2016/06/14 -- Addition of a __init__.py file in the server folder.
- Addition of a __init__.py file in the server folder.
- Correction of minor typos in the HTML shown.

0.1.0, 2016/06/13 -- Inclusion of two sample hidden services: hello-onion and headers-onion.
- Inclusion of two sample hidden services: hello-onion and headers-onion.
- Abstraction of the process of creating new services and making them available through Tor.
- Added the possibility of configuring where we want the hidden services to be run.
- Abstraction of the parameter configuration.

0.0.1, 2016/06/13 -- Initial commit.
- Initial commit.


server/__init__.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-
#
##################################################################################
#
#    Copyright 2016 Félix Brezo and Yaiza Rubio (i3visio, contacto@i3visio.com)
#
#    This program is free software. You can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
##################################################################################

__version__ = "0.1.1"


server/hello-onion.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-
#
##################################################################################
#
#    Copyright 2016 Félix Brezo and Yaiza Rubio (i3visio, contacto@i3visio.com)
#
#    This program is free software. You can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
##################################################################################


__NAME__ = "hello-onion"
__DESCRIPTION__ = "Sample .onion service."
__VERSION__ = "0.0.1"

# DEFAULT OPTIONS
DEFAULT_HOST = "localhost"
DEFAULT_PORT = "5000"
DEFAULT_DATA_DIR = "/tmp"

# Bottle imports
# --------------
from bottle import route, run, template
# To enable response content
from bottle import response
# To enable request processing...
from bottle import request
from bottle import redirect

@route('/')
def index():
    return '''
        <h1>¡Hola hola!</h1>
        Pues nada, que esto es un <i>hidden service</i> de prueba.
        Date una vuelta por:
        <ul>
        <li><a href="/javascript">/javascript</a>: para ver si tienes javascript activado</li>
        <li><a href="/hola/hacker">/hola/paco</a>: si te llamas Paco, para que te saludemos de forma personalizada</li>
        </ul>
    '''

@route('/javascript')
def sayHello():
    return '''
        Esta página tiene un Javascript chungo...
        <script>alert("Soy un javascript... Y soy malévolo... ¡MUAHAHA!");</script>
    '''

@route('/hola/<name>')
def sayHelloToName(name):
    print "Saludemos a " + name
    return template(
    '¡Hola <b>{{name}}</b>!', name=name)

"""
    Main execution of the server... It will generally won't need to be modified...
"""
if __name__ == "__main__":
    # Importing the service_manager that launches the application...
    import service_manager

    # Grabbing the parser
    args = service_manager.getParser(__NAME__, __DESCRIPTION__, DEFAULT_HOST, DEFAULT_PORT, DEFAULT_PORT, __VERSION__)

    # Calling the main function... We set the tor variable as True so as to make it available in the Tor Network....
    service_manager.main(name = __NAME__, host = args.host, port = args.port, data_dir = args.data_dir, debug = args.debug, clean = args.clean, tor=True)


server/headers-onion.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-
#
##################################################################################
#
#    Copyright 2016 Félix Brezo and Yaiza Rubio (i3visio, contacto@i3visio.com)
#
#    This program is free software. You can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
##################################################################################


__NAME__ = "headers-onion"
__DESCRIPTION__ = "Sample .onion service that returns the headers of a request."
__VERSION__ = "0.0.2"

# DEFAULT OPTIONS
DEFAULT_HOST = "localhost"
DEFAULT_PORT = "5001"
DEFAULT_DATA_DIR = "/tmp"

import os
import json

# Bottle imports
# --------------
from bottle import route, run, template
# To enable response content
from bottle import response
# To enable request processing...
from bottle import request
from bottle import redirect
# To handle API
from bottle import post, get, put, delete

def getFooter():
    return '''
        <br>
        <br>
        <small>Este <i>software</i> es <i>software</i> libre bajo licencia <a href="https://www.gnu.org/licenses/agpl.txt">AGPLv3</a>.</small>
    '''

@route('/')
def index():
    # get the Accept paramter from the HTTP Header
    print json.dumps(dict(request.headers),indent=2)
    return '''
        <h1>Bienvenido a esta huerta de cebollas</h1>

        <b>¡Hola hacker!</b>
        <br>
        Pues nada, que esto es un <i>hidden service</i> de prueba. Quizás quieras probar nuestra <i>aplicación</i> que te saluda de una forma muy especial.
        <br>
        Por ejemplo, prueba con <a href="/hola/paco">/hola/paco</a>, <a href="/hola/paco">/hola/james</a>, <a href="/hola/paco">/hola/ronaldo</a>...
    ''' + getFooter()

@route('/hola/<name>')
def sayHelloToName(name):
    cabeceras = json.dumps(dict(request.headers),indent=2)
    print cabeceras
    return template(
    ''' <b>¡Hola {{name}}!</b>
        Tu petición ha enviado las siguientes cabeceras:<br>
        <pre>{{headers}}</pre>
        ¿Ya te lo esperabas?'''
     + getFooter(), name=name, headers=cabeceras)

# Defining error websites...:
from bottle import error

@error(404)
def error404(error):
    print error
    return template('¿Pero qué buscas? ¡Aquí no hay ná-de-ná! <br> <pre>{{error}}</pre>', error=str(error))

@error(500)
def error500(error):
    print error
    return template('Error interno (código 500).. <br> <pre>{{error}}</pre>', error=str(error))


"""
    Main execution of the server... It will generally won't need to be modified...
"""
if __name__ == "__main__":
    # Importing the service_manager that launches the application...
    import service_manager

    # Grabbing the parser
    args = service_manager.getParser(__NAME__, __DESCRIPTION__, DEFAULT_HOST, DEFAULT_PORT, DEFAULT_PORT, __VERSION__)

    # Calling the main function... We set the tor variable as True so as to make it available in the Tor Network....
    service_manager.main(name = __NAME__, host = args.host, port = args.port, data_dir = args.data_dir, debug = args.debug, clean = args.clean, tor=True)


server/hidden_service_manager.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-
#
##################################################################################
#
#    Copyright 2016 Félix Brezo and Yaiza Rubio (i3visio, contacto@i3visio.com)
#
#    This program is free software. You can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
##################################################################################

import os
import shutil

# Bottle imports
# --------------
from bottle import run

# Stem controller imports
# -----------------------
from stem.control import Controller


"""
    Main function that launches the onion service
"""
def main(host = "localhost", port = 5000, data_dir = "/tmp", debug = True, clean = False, name = "a_hidden_service"):
    print(' * Connecting to tor...')

    with Controller.from_port(port=9051) as controller:
        controller.authenticate()

        # All hidden services have a directory on disk. Lets put ours in tor's data
        # directory.

        hidden_service_dir = os.path.join(controller.get_conf('DataDirectory', data_dir), name)

        # Create a hidden service where visitors of port 80 get redirected to local
        # port where Bottle will be listening.

        print(" * Creating our hidden service in %s..." % hidden_service_dir)
        try:
            result = controller.create_hidden_service(hidden_service_dir, 80, target_port = port)
            # The hostname is only available when we can read the hidden service
            # directory. This requires us to be running with the same user as tor.

            if result.hostname:
                print(" * Our service is available at %s. Press Ctrl+c to quit." % result.hostname)
            else:
                print(" * Unable to determine our service's hostname, probably due to being unable to read the hidden service directory.")

        except Exception as e:
            print(" * Something happened... Does it exist?")
            print(" * Trace:\n" + str(e))

        try:
            run(host=host, port=port, debug=debug)
        finally:
            # Shut down the hidden service and clean it off disk. Note that you *don't*
            # want to delete the hidden service directory if you'd like to have this
            # same *.onion address in the future.

            print(" * Shutting down our hidden service...")
            controller.remove_hidden_service(hidden_service_dir)
            if clean:
                print(" * Cleaning the hidden service directory...")
                shutil.rmtree(hidden_service_dir)


server/service_manager.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-
#
##################################################################################
#
#    Copyright 2016 Félix Brezo and Yaiza Rubio (i3visio, contacto@i3visio.com)
#
#    This program is free software. You can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
##################################################################################

import os
import shutil

# Bottle imports
# --------------
from bottle import run

# Stem controller imports
# -----------------------
from stem.control import Controller


"""
    Main function that launches the onion service
"""
def main(host = "localhost", port = 5000, data_dir = "/tmp", debug = True, clean = False, name = "a_hidden_service", tor=True):
    # If the user has decided that he/she wants to make the service available in the Tor Network...
    if tor:
        print(' * Connecting to tor...')

        with Controller.from_port(port=9051) as controller:
            controller.authenticate()

            # All hidden services have a directory on disk. Lets put ours in tor's data
            # directory.

            hidden_service_dir = os.path.join(controller.get_conf('DataDirectory', data_dir), name)

            # Create a hidden service where visitors of port 80 get redirected to local
            # port where Bottle will be listening.

            print(" * Creating our hidden service in %s..." % hidden_service_dir)
            try:
                result = controller.create_hidden_service(hidden_service_dir, 80, target_port = port)
                # The hostname is only available when we can read the hidden service
                # directory. This requires us to be running with the same user as tor.

                if result.hostname:
                    print(" * Our service is available at <%s>. Press Ctrl+c to quit." % result.hostname)
                else:
                    print(" * Unable to determine our service's hostname, probably due to being unable to read the hidden service directory.")

            except Exception as e:
                print(" * Something happened... Does it exist?")
                print(" * Trace:\n" + str(e))

            try:
                run(host=host, port=port, debug=debug)
            finally:
                # Shut down the hidden service and clean it off disk. Note that you *don't*
                # want to delete the hidden service directory if you'd like to have this
                # same *.onion address in the future.

                print(" * Shutting down our hidden service...")
                controller.remove_hidden_service(hidden_service_dir)
                if clean:
                    print(" * Cleaning the hidden service directory...")
                    shutil.rmtree(hidden_service_dir)
    # However, tje user may have decided to run this only in debug mode...
    else:
        try:
            run(host=host, port=port, debug=debug)
        finally:
            print(" * Shutting down our service...")



import argparse

def getParser(name, description, default_host, default_port, default_data_dir, version):
    # Grabbing the parser
    parser = argparse.ArgumentParser(description= name + " - " + description, prog = name, epilog='Check the README.md file for further details on the usage of this program or follow us on Twitter in <http://twitter.com/i3visio>.', add_help=False)

    # Configuring the processing options
    groupProcessing = parser.add_argument_group('Processing arguments', 'Configuring the way in which the server will behave.')
    groupProcessing.add_argument('-H', '--host', metavar='<host>',  required=False, default = default_host, action='store', help='defining the host in which the service will run. Default: .' + default_host)
    groupProcessing.add_argument('-P', '--port', metavar='<port>',  required=False, default = default_port, type=int, action='store', help='defining the port in which the service will run. Default: ' + str(default_host) + '.')
    groupProcessing.add_argument('-D', '--data_dir', metavar='<path>',  required=False, default = default_data_dir, action='store', help='defining the path of the data directory where the information will be stored. Default: ' + default_data_dir + '.')
    groupProcessing.add_argument('--debug',  required=False, default = False, action='store_true', help='setting debug mode. Default: False.')
    groupProcessing.add_argument('--clean',  required=False, default = False, action='store_true', help='setting Stem to clean the hidden service directory. Default: False.')

    # About options
    groupAbout = parser.add_argument_group('About arguments', 'Showing additional information about this program.')
    groupAbout.add_argument('-h', '--help', action='help', help='shows this help and exists.')
    groupAbout.add_argument('--license', required=False, action='store_true', default=False, help='shows the GPLv3+ license and exists.')
    groupAbout.add_argument('--version', action='version', version='%(prog)s ' +" " + version, help='shows the version of the program and exists.')

    args = parser.parse_args()

    return args
